include/deal.II/fe/fe_series.h

```diff
@@ -0,0 +1,391 @@
+   * functions \f$ \phi_{\bf k}({\bf x}) = \exp(2 \pi i\, {\bf k} \cdot {\bf x})\f$.
+   * For example in 1D the L2-orthogonality condition reads
+   * @f[
+   *   \int_0^1 \phi_k(x) \phi_l^\ast(x) dx=\delta_{kl}.
+   * @f]
+   * Note that \f$ \phi_{\bf k} = \phi_{-\bf k}^\ast \f$.
+   *
+   * The arbitrary scalar FE field on the reference element can be expanded in
+   * the complete orthogonal exponential basis as
+   * @f[
+   *    u({\bf x})
+   *    = \sum_{\bf k} c_{\bf k} \phi_{\bf k}({\bf x}).
+   * @f]
+   * From the orthogonality property of the basis, it follows that
+   * @f[
+   *    c_{\bf k} =
+   *    \int_{[0,1]^d} u({\bf x}) \phi_{\bf k}^\ast ({\bf x}) d{\bf x}\,.
+   * @f]
+   * It is this complex-valued expansion coefficients, that are calculated by
+   * this class. Note that \f$ u({\bf x}) = \sum_i u_i N_i({\bf x})\f$,
+   * where \f$ N_i({\bf x}) \f$ are real-valued FiniteElement shape functions.
+   * Consequently \f$ c_{\bf k} \equiv c_{-\bf k}^\ast \f$ and
+   * we only need to compute \f$ c_{\bf k} \f$ for positive indices
+   * \f$ \bf k \f$ .
+   *
+   * @author Denis Davydov, 2016.
+   */
+  template <int dim>
+  class Fourier : public Subscriptor
+  {
+  public:
+    /**
+     * A non-default constructor. The @p size_in_each_direction defines the number
+     * of modes in each direction, @p fe_collection is the hp::FECollection
+     * for which expansion will be used and @p q_collection is the hp::QCollection
+     * used to integrate the expansion for each FiniteElement
+     * in @p fe_collection.
+     */
+    Fourier(const unsigned int size_in_each_direction,
+            const hp::FECollection<dim> &fe_collection,
+            const hp::QCollection<dim> &q_collection);
+
+    /**
+     * Calculate @p fourier_coefficients of the cell vector field given by
+     * @p local_dof_values corresponding to FiniteElement with
+     * @p cell_active_fe_index .
+     */
+    void calculate(const dealii::Vector<double>     &local_dof_values,
+                   const unsigned int                cell_active_fe_index,
+                   Table<dim,std::complex<double> > &fourier_coefficients);
+
+  private:
+    /**
+     * hp::FECollection for which transformation matrices will be calculated.
+     */
+    const hp::FECollection<dim> &fe_collection;
+
+    /**
+     * hp::QCollection used in calculation of transformation matrices.
+     */
+    const hp::QCollection<dim>  &q_collection;
+
+    /**
+     * Ensure that the transformation matrix for FiniteElement index
+     * @p fe_index is calculated. If not, calculate it.
+     */
+    void ensure_existence(const unsigned int fe_index);
+
+    /**
+     * Angular frequencies \f$ 2 \pi {\bf k} \f$ .
+     */
+    Table<dim, Tensor<1,dim> > k_vectors;
+
+    /**
+     * Transformation matrices for each FiniteElement.
+     */
+    std::vector<std::shared_ptr< FullMatrix<std::complex<double> > > > fourier_transform_matrices;
```

Comment:
I do not believe that the `shared_ptr` is needed. I think that everything you do below with `ensure_existence` would work where you check if the size of the matrix is zero instead.